pkg/progressreader/broadcaster.go

```diff
@@ -0,0 +1,98 @@
+package progressreader
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// Broadcaster keeps track of one or more observers watching the progress
+// of an operation. For example, if multiple clients are trying to pull an
+// image, they share a Broadcaster for the download operation.
+type Broadcaster struct {
+	sync.RWMutex
+	// c is a channel that observers block on, waiting for the operation
+	// to finish.
+	c chan struct{}
+	// dataSignals is a list of channels which signal to the observers'
+	// receiveWrites goroutines that new data is available.
+	dataSignals []chan struct{}
+	// history is a buffer of the progress output so far, so a new observer
+	// can catch up.
+	history bytes.Buffer
+}
+
+// NewBroadcaster returns a Broadcaster structure
+func NewBroadcaster() *Broadcaster {
+	return &Broadcaster{
+		c:           make(chan struct{}),
+		dataSignals: []chan struct{}{},
+	}
+}
+
+// receiveWrites runs as a goroutine so that writes don't block the Write
+// function. It writes the new data in broadcaster.history each time there's
+// activity on the dataSignal channel.
+func (broadcaster *Broadcaster) receiveWrites(observer io.Writer, dataSignal <-chan struct{}) {
+	n := 0
+
+	for range dataSignal {
+		broadcaster.RLock()
+		newData := broadcaster.history.Bytes()[n:]
+		// Make a copy of newData so we can release the lock
+		sendData := make([]byte, len(newData), len(newData))
+		copy(sendData, newData)
+		broadcaster.RUnlock()
+
+		written, err := observer.Write(sendData)
+		if err != nil {
+			return
+		}
+		n += written
+	}
+}
+
+// Write adds data to the history buffer, and also writes it to all current
+// observers.
+func (broadcaster *Broadcaster) Write(p []byte) (n int, err error) {
+	broadcaster.Lock()
+	broadcaster.history.Write(p)
+	broadcaster.Unlock()
+
+	// Do a nonblocking write to each dataSignal to cause each writing
+	// goroutine to wake up.
+	for _, dataSignal := range broadcaster.dataSignals {
+		select {
+		case dataSignal <- struct{}{}:
```

Comment:
This might be a use case for [`sync.Cond`](http://golang.org/pkg/sync/#Cond), over channels.

In the `(*Broadcaster).Write`:

```
	broadcaster.Lock()
	defer broadcaster.Unlock()

	broadcaster.history.Write(p)
	broadcaster.cond.Broadcast()
```

In `(*Broadcaster). receiveWrites`:

```
for {
	broadcaster.cond.Wait()
	// ... catch up with write.

	// ... close detection
	select {
	case <-broadcast.c:
		return
	default:
		// noop: wait for write.
	}
}
```

Note that you can still do close detection using channels. Perhaps, we need a `(*Broadcaster).done`